src/_zkapauthorizer/tests/test_model.py

```diff
@@ -1153,3 +1162,45 @@ def test_roundtrip(self, pass_):
             Pass.from_bytes(pass_.pass_bytes),
             Equals(pass_),
         )
+
+
+class ReplicationTests(TestCase):
+    """
+    Tests for replication and recovery.
+    """
+
+    def test_recover(self):
+        """
+        Given a snapshot returned by ``snapshot`` can be loaded into an empty
+        ``VoucherStore`` using ``VoucherStore.call_if_empty`` with
+        ``StatefulRecoverer.recover``.
+        """
+        store = self.useFixture(
+            ConfiglessMemoryVoucherStore(
+                # Time is not relevant to this test
+                datetime.now,
+            )
+        ).store
+        # XXX Use VoucherStore.snapshot once it is available
+        snapshot_bytes = b"".join(statements_to_snapshot(snapshot(store._connection)))
+        downloader = make_canned_downloader(snapshot_bytes)
+        recoverer = StatefulRecoverer()
+
+        # StatefulRecoverer.recover should always succeed.  Verify that.
+        self.assertThat(
+            Deferred.fromCoroutine(
+                store.call_if_empty(partial(recoverer.recover, downloader))
+            ),
+            succeeded(Always()),
+        )
+
+        # It is not enough for StatefulRecoverer.recover to succeed, though.
+        # StatefulRecoverer must also end up in the succeeded state.
+        self.assertThat(
+            recoverer.state(),
+            Equals(
+                RecoveryState(
+                    stage=RecoveryStages.succeeded,
+                ),
+            ),
+        )
```

Comment:
Feels like this wants more (or maybe even a different test) about whether the database itself ends up in the right state, too? That is, a `recover` that just set the status to `succeeded` would pass this.

Reply:
I think this is covered by the `StatefulRecoverer` tests - `_zkapauthorizer.tests.test_recover.StatefulRecovererTests` (particular `test_state_recovered`) - which goes into more detail for the actual recovery code.

Reply:
I added some more words to the class docstring to try to describe its intended limited scope.